### PR TITLE
feat: add support for v-for by using ultrahtml to transform code

### PR DESCRIPTION
**Pull request by prashantpalikhe**

### CODE REVIEW

File: src/module.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@ import { name, version } from "../package.json";
 import autoLitWrapper from "./runtime/plugins/autoLitWrapper";
 
 export interface NuxtSsrLitOptions {
-  litElementPrefix: string | string[];
+  litElementPrefix: string[];
```
Comment (steveworkman) [inline src/module.ts:6]: This is a breaking change. Can we accommodate a single string?

File: src/runtime/plugins/autoLitWrapper.ts [hunk 2]
```diff
@@ -1,44 +1,77 @@
 import { pathToFileURL } from "node:url";
-import { parseQuery, parseURL } from "ufo";
-import MagicString from "magic-string";
+import { parseURL } from "ufo";
+import { parse, walk, ELEMENT_NODE, Node, render } from "ultrahtml";
+import { d } from "vue-bundle-renderer/dist/types-dfcc483f";
```
Comment (steveworkman) [inline src/runtime/plugins/autoLitWrapper.ts:4]: Where does this come from? Is it used?

File: src/runtime/plugins/autoLitWrapper.ts [hunk 3]
```diff
@@ -1,44 +1,77 @@
...
+      const ast = parse(code);
+
+      await walk(ast, (node) => {
+        if (node.type !== ELEMENT_NODE || !prefixRegex.test(node.name)) {
+          return;
+        }
+
+        const attributes = node.attributes;
+
+        const hasVFor = node.attributes["v-for"];
```
Comment (steveworkman) [inline src/runtime/plugins/autoLitWrapper.ts:42]:
```suggestion
        const hasVFor = node.attributes[V_FOR_ATTRIBUTE];
```

**Review by steveworkman — CHANGES_REQUESTED** [review]

File: src/runtime/plugins/autoLitWrapper.ts [hunk 4]
```diff
@@ -1,44 +1,77 @@
...
+
+          delete node.attributes[V_FOR_ATTRIBUTE];
+          delete node.attributes[V_FOR_KEY_ATTRIBUTE];
+        }
+
+        Object.keys(node.attributes).forEach((key) => {
+          if (key === "") {
+            delete node.attributes[key];
+          }
+        });
```
Comment (flauwekeul) [inline src/runtime/plugins/autoLitWrapper.ts:65]: Not sure if this works, but it's simpler:
```suggestion
        delete node.attributes[""];
```

File: src/runtime/plugins/autoLitWrapper.ts [hunk 5]
```diff
@@ -1,44 +1,77 @@
...
-      const endTagRegex = new RegExp(`<\\/(${prefix}[a-z-]+)>`, "g");
+      const transformedCode = await render(ast);
 
-      s.replace(openTagRegex, "<LitWrapper><$1").replace(endTagRegex, "</$1></LitWrapper>");
-      // No need to check for s.hasChanged() - nuxt warns about sourcemap potentially being wrong if it's not re-generated
       return {
-        code: s.toString(),
-        map: sourcemap.server ? s.generateMap({ source: id, includeContent: true }) : undefined
+        code: transformedCode,
+        map: null
```
Comment (flauwekeul) [inline src/runtime/plugins/autoLitWrapper.ts:74]: Why not remove `map`?

**Review by flauwekeul — CHANGES_REQUESTED** [review]
I see this must've been a fun code change :)

File: src/runtime/plugins/autoLitWrapper.ts (hunk 2 above)
Comment (prashantpalikhe) [reply src/runtime/plugins/autoLitWrapper.ts:4]: No clue. This gotta go!

File: src/module.ts (hunk 1 above)
Comment (prashantpalikhe) [reply src/module.ts]: Done!

File: src/runtime/plugins/autoLitWrapper.ts (hunk 2 above)
Comment (prashantpalikhe) [reply src/runtime/plugins/autoLitWrapper.ts]: Done!

File: src/runtime/plugins/autoLitWrapper.ts (hunk 3 above)
Comment (prashantpalikhe) [reply src/runtime/plugins/autoLitWrapper.ts]: Done!

File: src/runtime/plugins/autoLitWrapper.ts [hunk 6]
```diff
@@ -1,44 +1,77 @@
...
+        return;
+      }
 
-      if (!isVueFile) {
+      if (!litElementPrefix.some((prefix) => code.includes(`<${prefix}`))) {
         return;
       }
 
-      const prefix = Array.isArray(litElementPrefix) ? `(${litElementPrefix.join("|")})` : litElementPrefix;
+      const prefixRegex = new RegExp(`^(${litElementPrefix.join("|")})`, "i");
```
[the remark this answers is not in the document]
Comment (prashantpalikhe) [reply src/runtime/plugins/autoLitWrapper.ts]: The regex is built with the litElementPrefix array passed as an argument to the function. So I don't think we can move this out.

File: src/runtime/plugins/autoLitWrapper.ts (hunk 4 above)
Comment (prashantpalikhe) [reply src/runtime/plugins/autoLitWrapper.ts]: Done!

File: src/runtime/plugins/autoLitWrapper.ts (hunk 5 above)
Comment (prashantpalikhe) [reply src/runtime/plugins/autoLitWrapper.ts]: Done!

File: src/runtime/plugins/autoLitWrapper.ts [hunk 7]
```diff
@@ -1,44 +1,77 @@
...
-}: AutoLitWrapperOptions) {
+export default function autoLitWrapper({ litElementPrefix = [] }: NuxtSsrLitOptions) {
   return {
     name: "autoLitWrapper",
-    transform(code: string, id: string) {
-      const { pathname, search } = parseURL(decodeURIComponent(pathToFileURL(id).href));
-      const { type } = parseQuery(search);
+    async transform(code: string, id: string) {
+      const { pathname } = parseURL(decodeURIComponent(pathToFileURL(id).href));
+      const isVueFile = pathname.endsWith(".vue");
```
[the remark this answers is not in the document]
Comment (prashantpalikhe) [reply src/runtime/plugins/autoLitWrapper.ts:16]: Will have to think about that. Will do that on a separate PR.

File: src/runtime/plugins/autoLitWrapper.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (flauwekeul) [reply src/runtime/plugins/autoLitWrapper.ts]: I meant move it out from the `transform` function, but inside the `autoLitWrapper` function.

File: src/runtime/plugins/autoLitWrapper.ts [hunk 8]
```diff
@@ -1,44 +1,74 @@
...
-  sourcemap = { client: false, server: true }
-}: AutoLitWrapperOptions) {
+export default function autoLitWrapper({ litElementPrefix = [] }: NuxtSsrLitOptions) {
   return {
     name: "autoLitWrapper",
-    transform(code: string, id: string) {
-      const { pathname, search } = parseURL(decodeURIComponent(pathToFileURL(id).href));
-      const { type } = parseQuery(search);
+    async transform(code: string, id: string) {
+      const litElementPrefixes = Array.isArray(litElementPrefix) ? litElementPrefix : [litElementPrefix];
```
Comment (flauwekeul) [inline src/runtime/plugins/autoLitWrapper.ts:13]:
```suggestion
      const litElementPrefixes = Array.from(litElementPrefix);
```

**Review by flauwekeul — APPROVED** [review]

**Pull request merged**